R/mediate.R

```diff
@@ -389,48 +409,52 @@ setMethod("getMediation", signature = c(x = "SummarizedExperiment"),
         treat = "Treatment", mediator = "Mediator",
         covariates = covariates, ...
     )
-    return(med_out)
-}
-
 
-# Update list of results
-.update.results <- function(results, med_out, mediator) {
-    # Update model variables
-    results[["Mediator"]] <- c(results[["Mediator"]], mediator)
-    # Update stats of ACME (average causal mediation effect)
-    results[["ACME_estimate"]] <- c(results[["ACME_estimate"]], med_out$d.avg)
-    results[["ACME_pval"]] <- c(results[["ACME_pval"]], med_out$d.avg.p)
-    # Update stats of ADE (average direct effect)
-    results[["ADE_estimate"]] <- c(results[["ADE_estimate"]], med_out$z.avg)
-    results[["ADE_pval"]] <- c(results[["ADE_pval"]], med_out$z.avg.p)
-    # Add current model to metadata
-    results[["Model"]][[length(results[["Model"]]) + 1]] <- med_out
-    return(results)
+    return(med_out)
 }
 
 # Combine results into output dataframe
-.make.output <- function(results, p.adj.method, add.metadata) {
-    
-    # Create dataframe with model variables, effect sizes and p-values
-    med_df <- do.call(data.frame, results[seq_len(length(results) - 1)])
-    
+#' @importFrom dplyr select
+#' @importFrom tidyr function starts_with ends_with unnest_wider
+#' @importFrom stringr str_extract str_replace_all
+.make_output <- function(models, p.adj.method, add.metadata, sort) {
+    # Combine results
+    res <- do.call(rbind, models) |> as.data.frame()
+    # Select certain data types
+    res <- res |>
+        select(tidyr::starts_with(c("d.avg", "z.avg", "tau"))) |>
+        select(-tidyr::ends_with(c("sims")))
+    # Get columns with scalar values and turn them into vector instead of list
+    cols <- vapply(res, function(col) all(lengths(col) == 1L), logical(1L))
+    res[, cols] <- lapply(res[, cols], unlist)
+    # Unnest confidence interval columns
+    res <- res |> unnest_wider(tidyr::ends_with(".ci"), names_sep = "_")
+    # Replace confidence interval values with "lower" and "upper"
+    limits <- unique(str_extract(colnames(res), "([0-9.]+)%"))
+    limits <- limits[!is.na(limits)]
+    limits <- limits[order(as.numeric(gsub("%", "", limits)))]
+    lookup <- c("lower", "upper")
+    names(lookup) <- limits
+    colnames(res) <- str_replace_all(colnames(res), lookup)
+    # Tidy other names also
+    lookup <- c("d.avg" = "ACME", "z.avg" = "ADE", "tau" = "Total",
+        "\\.p" = "_pval", "\\.ci_" = "_")
+    colnames(res) <- str_replace_all(colnames(res), lookup)
```

Comment:
I modified this little bit. You can still change renaming as it might be over-engineered, although it should be robust